src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -0,0 +1,179 @@
+                    "warning" -> HighlightSeverity.WEAK_WARNING
+                    else -> HighlightSeverity.INFORMATION
+                }
+
+            // If spans are empty we add a "global" error
+            if (message.spans.isEmpty()) {
+                if (topMessage.target.src_path == filePath) {
+                    // add a global annotation
+                    val annotation = holder.createAnnotation(severity, TextRange.EMPTY_RANGE, message.message)
+                    annotation.isFileLevelAnnotation = true
+                    annotation.setNeedsUpdateOnTyping(true)
+                    annotation.tooltip = escapeHtml(message.message) + " " + message.code.formatAsLink().orEmpty()
+                }
+            } else {
+                val problemGroup = ProblemGroup { message.message }
+
+                val primarySpan =
+                    message.spans
+                        .filter { it.is_primary }
+                        .filter { isValidSpan(it) }
+                        .filter { filePath.endsWith(PathUtil.getCanonicalPath(it.file_name)) }
+                        .firstOrNull()
+
+                if (primarySpan != null) {
+                    val annotation = createAnnotation(primarySpan, message, severity, doc, holder)
+                    annotation.problemGroup = problemGroup
+                    annotation.setNeedsUpdateOnTyping(true)
+                }
+            }
+        }
+    }
+
+    fun createAnnotation(span: CargoSpan, message: CargoMessage, severity: HighlightSeverity, doc: Document,
+                         holder: AnnotationHolder): Annotation {
+
+        fun toOffset(line: Int, column: Int): Int {
+            val lineStart = doc.getLineStartOffset(line)
+            return lineStart + column
+        }
+
+        // The compiler message lines and columns are 1 based while intellij idea are 0 based
+        val textRange =
+            TextRange(
+                toOffset(span.line_start - 1, span.column_start - 1),
+                toOffset(span.line_end - 1, span.column_end - 1))
+
+        val tooltip = run {
+            val lines = ArrayList<String>()
+            val code = message.code.formatAsLink()
+            lines.add(escapeHtml(message.message) + if (code == null) "" else " $code")
+
+            if (span.label != null && !message.message.startsWith(span.label))
+                lines.add(escapeHtml(span.label))
+
+            message.children
+                .filter { !it.message.isBlank() }
+                .forEach {
+                    val prefix = when (it.level) {
+                        "note" -> "Note: "
+                        "help" -> "Help: "
+                        else -> ""
+                    }
+                    lines.add(prefix + escapeHtml(it.message))
+                }
+
+            lines.joinToString("<br>").replace("\n", "<br>")
+        }
+
+        return holder.createAnnotation(severity, textRange, message.message, "<html>$tooltip</html>")
```

Comment:
I cannot reproduce this tooltip:

![image](https://cloud.githubusercontent.com/assets/873919/26793347/54f34ad2-4a26-11e7-8984-2b4fb4bf3f32.png)

Could you provide full repro snippet?